File: apps/android/app/src/test/kotlin/com/loonext/android/core/jobs/WorkPhaseTest.kt
package com.loonext.android.core.jobs

import java.io.File
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * #294 — before and after, and that this phone groups a job the way the laptop does.
 *
 * A drift here means one client shows a customer four visits and another shows one
 * flat pile of photos. It would not fail to compile, which is why the port gets its
 * own test.
 */
class WorkPhaseTest {

    private data class Photo(
        override val id: String,
        override val noteId: String? = null,
        override val workPhase: String? = null,
        override val addedByUserId: String? = null,
        override val createdAt: String = "2026-08-08T10:00:00Z",
    ) : JobPhotoLike

    // ------------------------------------------------------------ the two labels

    @Test
    fun `it is before and after, and nothing else`() {
        // "During" is a category somebody invented in a meeting. A tech takes a
        // handful when they arrive and a handful when they finish.
        assertEquals(listOf("before", "after"), WorkPhase.ALL)
        assertEquals("Before", WorkPhase.label("before"))
        assertEquals("After", WorkPhase.label("after"))
    }

    @Test
    fun `anything else is not a phase`() {
        assertTrue(WorkPhase.isPhase("before"))
        assertTrue(WorkPhase.isPhase("after"))
        for (bad in listOf("during", "BEFORE", "", null)) {
            assertTrue(bad ?: "null", !WorkPhase.isPhase(bad))
        }
    }

    // --------------------------------------------------------------- the grouping

    @Test
    fun `each note's files go together`() {
        val groups = groupJobPhotos(
            listOf(
                Photo("a", noteId = "n1", createdAt = "2026-08-08T09:00:00Z"),
                Photo("b", noteId = "n2", createdAt = "2026-08-08T15:00:00Z"),
                Photo("c", noteId = "n1", createdAt = "2026-08-08T09:00:05Z"),
            ),
        )
        assertEquals(2, groups.size)
        assertEquals(listOf("a", "c"), groups[0].items.map { it.id })
        assertEquals(listOf("b"), groups[1].items.map { it.id })
    }

    @Test
    fun `two notes written in the same second stay apart`() {
        // THE CASE THAT MATTERS for keying on the note rather than on the time: two
        // visits merged into one is a job record that says something untrue.
        val same = "2026-08-08T09:00:00Z"
        val groups = groupJobPhotos(
            listOf(
                Photo("a", noteId = "n1", createdAt = same),
                Photo("b", noteId = "n2", createdAt = same),
            ),
        )
        assertEquals(2, groups.size)
    }

    @Test
    fun `everything the customer texted is one group`() {
        val groups = groupJobPhotos(
            listOf(
                Photo("a", noteId = null, createdAt = "2026-08-08T08:00:00Z"),
                Photo("b", noteId = null, createdAt = "2026-08-08T08:00:01Z"),
                Photo("c", noteId = "n1", createdAt = "2026-08-08T09:00:00Z"),
            ),
        )
        assertEquals(2, groups.size)
        assertNull(groups[0].noteId)
        assertEquals(listOf("a", "b"), groups[0].items.map { it.id })
    }

    @Test
    fun `a visit is ordered by when it started, not by its last upload`() {
        val groups = groupJobPhotos(
            listOf(
                Photo("late", noteId = "n1", createdAt = "2026-08-08T18:00:00Z"),
                Photo("early", noteId = "n1", createdAt = "2026-08-08T08:00:00Z"),
                Photo("midday", noteId = "n2", createdAt = "2026-08-08T12:00:00Z"),
            ),
        )
        assertEquals(listOf("n1", "n2"), groups.map { it.noteId })
        assertEquals("2026-08-08T08:00:00Z", groups[0].at)
    }

    @Test
    fun `the label and the author come from the note`() {
        val groups = groupJobPhotos(
            listOf(Photo("a", noteId = "n1", workPhase = "after", addedByUserId = "u1")),
        )
        assertEquals("after", groups[0].workPhase)
        assertEquals("u1", groups[0].addedByUserId)
    }

    @Test
    fun `nothing in, nothing out`() {
        assertEquals(emptyList<JobPhotoGroup<Photo>>(), groupJobPhotos(emptyList<Photo>()))
    }

    // ---------------------------------------------------------------- the summary

    @Test
    fun `the summary counts each label`() {
        assertEquals(
            "2 before, 1 after",
            jobPhaseSummary(
                listOf(
                    Photo("a", workPhase = "before"),
                    Photo("b", workPhase = "before"),
                    Photo("c", workPhase = "after"),
                ),
            ),
        )
    }

    @Test
    fun `an unlabelled job gets no summary at all`() {
        // Not "0 before, 0 after", which reads as a broken count rather than as a
        // job whose photos nobody classified — and most jobs will be that.
        assertNull(jobPhaseSummary(listOf(Photo("a"), Photo("b"))))
        assertNull(jobPhaseSummary(emptyList()))
    }

    // ------------------------------------------------------- against the original

    /** The shared source, with carriage returns stripped — this tree is CRLF. */
    private fun repoFile(relative: String): String {
        var dir: File? = File("").absoluteFile
        while (dir != null) {
            val candidate = File(dir, relative)
            if (candidate.exists()) return candidate.readText().filterNot { it == '\r' }
            dir = dir.parentFile
        }
        throw AssertionError("$relative not found from ${File("").absolutePath}")
    }

    @Test
    fun `the labels match the shared module`() {
        // #228: the WORDS live in the web catalogue now. The shared module
        // names keys, and the phase ids ("before"/"after") — which are wire
        // values a photo carries — are still asserted against it below.
        //
        // Sliced to the English half: the French holds the same keys, and a
        // `contains` over the whole file would ask whether a label appears in
        // EITHER language.
        val shared = repoFile("apps/web/src/i18n/sections/domain.ts")
            .substringAfter("export const domainEn")
            .substringBefore("export const domainFr")
        for (label in listOf(
            WorkPhase.label("before"),
            WorkPhase.label("after"),
            WorkPhase.UNSET_LABEL,
        )) {
            assertTrue("this label has drifted: $label", shared.contains("\"$label\""))
        }
        assertTrue(
            "the hint has drifted: ${WorkPhase.HINT}",
            shared.replace(Regex("\\s+"), " ").contains(
                WorkPhase.HINT.replace(Regex("\\s+"), " "),
            ),
        )
    }

    @Test
    fun `the shared module still knows only these two phases`() {
        // A third phase added there and not here would leave this phone unable to
        // draw a label the server accepts.
        val shared = repoFile("packages/shared/src/work-phase.ts")
        val declared = Regex("""WORK_PHASES = \[([^\]]*)\]""")
            .find(shared)
            ?.groupValues
            ?.get(1)
            ?: throw AssertionError("WORK_PHASES has moved — point this test at it")
        val names = Regex("\"([a-z]+)\"").findAll(declared).map { it.groupValues[1] }.toList()
        assertEquals(WorkPhase.ALL, names)
    }
}

File: packages/shared/src/work-phase.test.ts
import { describe, expect, it } from "vitest";

import {
  WORK_PHASES,
  WORK_PHASE_LABELS,
  WORK_PHASE_UNSET_LABEL,
  groupJobPhotos,
  isWorkPhase,
  jobPhaseSummary,
  type JobPhotoLike,
} from "./work-phase";

import { EN as WEB_EN, FR_CA as WEB_FR } from "../../../apps/web/src/i18n/catalog";

/** #228 — the module names keys now, so the copy assertions resolve them. */
function look(table: unknown, key: string): string {
  const [section, name] = key.split(".");
  const value = (table as Record<string, Record<string, string>>)[section]?.[
    name
  ];
  if (typeof value !== "string") throw new Error(`no entry for ${key}`);
  return value;
}

function photo(over: Partial<JobPhotoLike> & { id: string }): JobPhotoLike {
  return { created_at: "2026-08-08T10:00:00Z", ...over };
}

describe("the two labels (#294)", () => {
  it("is before and after, and nothing else", () => {
    // The one classification the trade uses. "During" is a category somebody
    // invented in a meeting; a tech takes a handful when they arrive and a handful
    // when they finish.
    // The ids, which are wire values a photo carries and did not move.
    expect([...WORK_PHASES]).toEqual(["before", "after"]);
    // The words, where the words live since #228.
    expect(look(WEB_EN, WORK_PHASE_LABELS.before)).toBe("Before");
    expect(look(WEB_EN, WORK_PHASE_LABELS.after)).toBe("After");
    expect(look(WEB_FR, WORK_PHASE_LABELS.before)).toBe("Avant");
    expect(look(WEB_FR, WORK_PHASE_LABELS.after)).toBe("Après");
  });

  it("#228: does not offer 'None' for the third choice, in either language", () => {
    // The unset label is named rather than "None" on purpose: most notes are
    // neither a before nor an after, and "None" invites a tech to think they
    // have failed to fill something in. "Aucun" is the French version of that
    // same mistake.
    expect(look(WEB_EN, WORK_PHASE_UNSET_LABEL).toLowerCase()).not.toBe("none");
    const fr = look(WEB_FR, WORK_PHASE_UNSET_LABEL).toLowerCase();
    expect(fr).not.toBe("aucun");
    expect(fr).not.toBe("aucune");
    // It says what it IS instead — neither one nor the other.
    expect(fr).toContain("avant");
    expect(fr).toContain("après");
  });

  it("refuses anything that is not one of them", () => {
    expect(isWorkPhase("before")).toBe(true);
    expect(isWorkPhase("after")).toBe(true);
    for (const bad of ["during", "BEFORE", "", null, undefined, 1, {}]) {
      expect(isWorkPhase(bad), String(bad)).toBe(false);
    }
  });
});

describe("grouping a job's photos into visits (#294)", () => {
  it("puts each note's files together", () => {
    const groups = groupJobPhotos([
      photo({ id: "a", note_id: "n1", created_at: "2026-08-08T09:00:00Z" }),
      photo({ id: "b", note_id: "n2", created_at: "2026-08-08T15:00:00Z" }),
      photo({ id: "c", note_id: "n1", created_at: "2026-08-08T09:00:05Z" }),
    ]);
    expect(groups).toHaveLength(2);
    expect(groups[0].items.map((i) => i.id)).toEqual(["a", "c"]);
    expect(groups[1].items.map((i) => i.id)).toEqual(["b"]);
  });

  it("keeps two notes written in the same second apart", () => {
    // THE CASE THAT MATTERS for keying on the note rather than on the time: two
    // visits' worth of photos merged into one is a job record that says something
    // untrue about the day.
    const same = "2026-08-08T09:00:00Z";
    const groups = groupJobPhotos([
      photo({ id: "a", note_id: "n1", created_at: same }),
      photo({ id: "b", note_id: "n2", created_at: same }),
    ]);
    expect(groups).toHaveLength(2);
  });

  it("collects everything the customer texted into one group", () => {
    // It did not arrive in visits, and splitting it into per-file groups would
    // invent structure that is not there.
    const groups = groupJobPhotos([
      photo({ id: "a", note_id: null, created_at: "2026-08-08T08:00:00Z" }),
      photo({ id: "b", note_id: null, created_at: "2026-08-08T08:00:01Z" }),
      photo({ id: "c", note_id: "n1", created_at: "2026-08-08T09:00:00Z" }),
    ]);
    expect(groups).toHaveLength(2);
    expect(groups[0].note_id).toBeNull();
    expect(groups[0].items.map((i) => i.id)).toEqual(["a", "b"]);
  });

  it("orders visits by when they started, not by when the last file landed", () => {
    // A slow second upload must not move a visit later in the day than it
    // happened. The group's time is its EARLIEST file.
    const groups = groupJobPhotos([
      photo({ id: "late-of-early", note_id: "n1", created_at: "2026-08-08T18:00:00Z" }),
      photo({ id: "early-of-early", note_id: "n1", created_at: "2026-08-08T08:00:00Z" }),
      photo({ id: "midday", note_id: "n2", created_at: "2026-08-08T12:00:00Z" }),
    ]);
    expect(groups.map((g) => g.note_id)).toEqual(["n1", "n2"]);
    expect(groups[0].at).toBe("2026-08-08T08:00:00Z");
  });

  it("carries the label and the author from the note", () => {
    const groups = groupJobPhotos([
      photo({
        id: "a",
        note_id: "n1",
        work_phase: "after",
        added_by_user_id: "u1",
      }),
    ]);
    expect(groups[0].work_phase).toBe("after");
    expect(groups[0].added_by_user_id).toBe("u1");
  });

  it("is stable, so nothing shuffles between two renders", () => {
    const items = [
      photo({ id: "a", note_id: "n1", created_at: "2026-08-08T09:00:00Z" }),
      photo({ id: "b", note_id: "n2", created_at: "2026-08-08T09:00:00Z" }),
      photo({ id: "c", note_id: "n1", created_at: "2026-08-08T09:00:00Z" }),
    ];
    const once = groupJobPhotos(items).map((g) => g.note_id);
    const twice = groupJobPhotos([...items].reverse()).map((g) => g.note_id);
    expect(once).toEqual(twice);
  });

  it("returns nothing for nothing", () => {
    expect(groupJobPhotos([])).toEqual([]);
  });
});

describe("the one-line summary (#294)", () => {
  it("counts each label", () => {
    expect(
      jobPhaseSummary([
        photo({ id: "a", work_phase: "before" }),
        photo({ id: "b", work_phase: "before" }),
        photo({ id: "c", work_phase: "after" }),
      ]),
    ).toBe("2 before, 1 after");
  });

  it("names only the label that is there", () => {
    expect(jobPhaseSummary([photo({ id: "a", work_phase: "before" })])).toBe(
      "1 before",
    );
  });

  it("says nothing at all when nothing is labelled", () => {
    // Not "0 before, 0 after", which reads as a broken count rather than as a job
    // whose photos nobody classified — and most jobs will be that.
    expect(jobPhaseSummary([photo({ id: "a" }), photo({ id: "b" })])).toBeNull();
    expect(jobPhaseSummary([])).toBeNull();
  });
});
